grabbit/src/main/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessor.groovy

```diff
@@ -88,10 +90,51 @@ class JcrNodesProcessor implements ItemProcessor<JcrNode, Node> {
             }
         }
 
-        final List<JcrProperty> properties = jcrNode.properties.toList()
-        Node.Builder nodeBuilder = Node.newBuilder()
+        // It checks if the node parent has required primary type as "nt:hierarchyNode"
+        // and then skip it as the node has already been processed by parent node
+        if(isRequiredHierarchyChildNode(jcrNode)) {
+            return null;
+        } else if(isHierarchyNode(jcrNode)) {
+            // Parent "nt:hierarchyNode" type will be processed
+            return buildNode(jcrNode, getRequiredChildNodes(jcrNode))
+        } else {
+            // Non-Hierarchy node logic
+            return buildNode(jcrNode, null)
+        }
+    }
+
+    private Node buildNode(JcrNode jcrNode, Collection<JcrNode> childNodes) {
+        final Node.Builder nodeBuilder = Node.newBuilder()
         nodeBuilder.setName(jcrNode.path)
+        nodeBuilder.setProperties(buildProperties(jcrNode))
+        childNodes?.each {
+            nodeBuilder.addChildNode(buildNode(it, getRequiredChildNodes(it)))
```

Comment:
The recursive algorithm goes up to the node till it will keep finding `nt:hierarchy parent + required child node` so even there will be more child nodes but if it does not satisfy the same then those will be not processed.